pkg/services/ngalert/store/instance_database.go

```diff
@@ -88,43 +88,84 @@ func (st DBstore) ListAlertInstances(ctx context.Context, cmd *models.ListAlertI
+		keysPerRow := 9
+		maxRows := 20
+
+		// Prepare the statement to bind each argument to.
+		bigUpsertSQL, err := st.SQLStore.Dialect.UpsertMultipleSQL(
+			"alert_instance",
+			[]string{"rule_org_id", "rule_uid", "labels_hash"},
+			[]string{"rule_org_id", "rule_uid", "labels", "labels_hash", "current_state", "current_reason", "current_state_since", "current_state_end", "last_eval_time"},
+			maxRows)
 		if err != nil {
 			return err
 		}
 
-		alertInstance := &models.AlertInstance{
-			RuleOrgID:         cmd.RuleOrgID,
-			RuleUID:           cmd.RuleUID,
-			Labels:            cmd.Labels,
-			LabelsHash:        labelsHash,
-			CurrentState:      cmd.State,
-			CurrentReason:     cmd.StateReason,
-			CurrentStateSince: cmd.CurrentStateSince,
-			CurrentStateEnd:   cmd.CurrentStateEnd,
-			LastEvalTime:      cmd.LastEvalTime,
+		bigStmt, err := sess.DB().Prepare(bigUpsertSQL)
+		if err != nil {
+			return err
 		}
 
-		if err := models.ValidateAlertInstance(alertInstance); err != nil {
-			return err
+		maxArgs := maxRows * keysPerRow
+		args := make([]interface{}, 0, maxArgs)
+		for _, alertInstance := range cmd {
+			if len(args) >= maxArgs {
+				_, err = bigStmt.ExecContext(ctx, args...)
+				if err != nil {
+					return err
+				}
+				args = args[:0]
+			}
+
+			labelTupleJSON, err := alertInstance.Labels.StringKey()
+			if err != nil {
+				return err
+			}
+
+			if err := models.ValidateAlertInstance(alertInstance); err != nil {
+				return err
+			}
+
+			args = append(args,
+				alertInstance.RuleOrgID, alertInstance.RuleUID, labelTupleJSON, alertInstance.LabelsHash,
+				alertInstance.CurrentState, alertInstance.CurrentReason, alertInstance.CurrentStateSince.Unix(),
+				alertInstance.CurrentStateEnd.Unix(), alertInstance.LastEvalTime.Unix())
 		}
 
-		params := append(make([]interface{}, 0), alertInstance.RuleOrgID, alertInstance.RuleUID, labelTupleJSON, alertInstance.LabelsHash, alertInstance.CurrentState, alertInstance.CurrentReason, alertInstance.CurrentStateSince.Unix(), alertInstance.CurrentStateEnd.Unix(), alertInstance.LastEvalTime.Unix())
+		if len(args) >= 0 {
+			upsertSQL, err := st.SQLStore.Dialect.UpsertMultipleSQL(
+				"alert_instance",
+				[]string{"rule_org_id", "rule_uid", "labels_hash"},
+				[]string{"rule_org_id", "rule_uid", "labels", "labels_hash", "current_state", "current_reason", "current_state_since", "current_state_end", "last_eval_time"},
+				len(args)/keysPerRow)
+			if err != nil {
+				return err
+			}
 
-		upsertSQL := st.SQLStore.Dialect.UpsertSQL(
-			"alert_instance",
-			[]string{"rule_org_id", "rule_uid", "labels_hash"},
-			[]string{"rule_org_id", "rule_uid", "labels", "labels_hash", "current_state", "current_reason", "current_state_since", "current_state_end", "last_eval_time"})
-		_, err = sess.SQL(upsertSQL, params...).Query()
-		if err != nil {
-			return err
+			stmt, err := sess.DB().Prepare(upsertSQL)
```

Comment:
The prepared statements look the same, can't we prepare it once and then use it for each upsert?